File: src/llm_trajectory/transpiler/transpiler.py
from __future__ import annotations

from dataclasses import dataclass

from llm_trajectory.primitives.registry import get_handler
from llm_trajectory.scene.models import Scene
from llm_trajectory.transpiler.trajectory import Trajectory


@dataclass
class ToolCall:
    name: str
    params: dict


class Transpiler:
    def transpile(self, tool_calls: list[ToolCall], scene: Scene) -> Trajectory:
        trajectory = Trajectory()
        for call in tool_calls:
            handler = get_handler(call.name)
            waypoint = handler(scene, **call.params)
            trajectory.add(waypoint)
        return trajectory

File: frontend/src/core/transpiler.ts
import type { Scene } from "./scene";
import type { ToolCall, Trajectory, Waypoint } from "./types";
import { getHandler } from "./primitives";

export function transpile(
  toolCalls: readonly ToolCall[],
  scene: Scene,
): Trajectory {
  const waypoints: Waypoint[] = toolCalls.map((call) => {
    const prim = getHandler(call.name);
    return prim.handler(scene, call.params);
  });

  return { metadata: {}, waypoints };
}
